### PR TITLE
[core] Cache runtime env core worker

### PR DESCRIPTION
**Pull request by dentiny**
Followup PR for https://github.com/ray-project/ray/pull/48749
This PR caches json and protobuf deserialization for serialized runtime env.
Checked with benchmark (https://github.com/ray-project/ray/blob/master/release/benchmarks/distributed/test_many_tasks.py), verified throughput improves from ~11550 to ~ 13565, which is ~17%.

### CODE REVIEW

File: src/ray/core_worker/core_worker.h
Comment (jjyao) [inline src/ray/core_worker/core_worker.h:1865]: This is needed to merge this PR otherwise we will have unbounded memory usage.

**Review by jjyao — COMMENTED** [review]
Is the 17% improvement observed on the CI release test or local laptop?

**Comment by dentiny** [comment]
> Is the 17% improvement observed on the CI release test or local laptop?

From local

File: src/ray/core_worker/core_worker.h
Comment (dentiny) [reply src/ray/core_worker/core_worker.h:1865]: OK, let me make another PR for LRU cache under `util` folder

Comment (dentiny) [reply src/ray/core_worker/core_worker.h]: I found there's an existing implementation in boost, updated implementation.

File: src/ray/core_worker/core_worker.cc
Comment (rynewang) [inline src/ray/core_worker/core_worker.cc:2227]: some thoughts:

1. `parent` is const per job, so instead of parsing everytime (master) or asking a cache every time (this PR), we can just define a new, per-job `WorkerContext::GetCurrentJobRuntimeEnv()`, just like the per-task `WorkerContext::GetCurrentRuntimeEnv()`, which parses ahead of the time and it kept throughout the job.
2. same thing to the copy of parent_runtime_env_info
3. the cache in this PR only caches the json/pb parsing part, but now the runtime env merging part. Considering we always use the merged rtenv, we can do caching in a job level for merged rtenvs. How it works: make a LRU cache for std::shared_ptr<rpc::RuntimeEnvInfo>; on job assignment clear it; on call to OverrideTaskOrActorRuntimeEnvInfo, first consult the LRU cache.

Comment (rynewang) [reply src/ray/core_worker/core_worker.cc:2227]: we can also discuss this offline

File: src/ray/core_worker/core_worker.h
Comment (jjyao) [inline src/ray/core_worker/core_worker.h:1865]: Can we temporarily change this back to std::unordered_map so that this PR compiles and we can see the real perf impact on the release test. I'm curious whether the release test will see the same perf improvement as what you observed locally.

File: src/ray/core_worker/core_worker.cc
Comment (dentiny) [reply src/ray/core_worker/core_worker.cc:2227]: Ok, it seems there're two things I need to cache:
- Job-wise parsed json
- Merged pb runtime env info, key-ed by serialized runtime env

Comment (dentiny) [reply src/ray/core_worker/core_worker.cc:2226]: I think the overall proposal makes sense to me, but it *might* adds runtime overhead due to the enlarged scoped of cached value execution.
The ideal implementation might be having another sharded cache to reduce critical section, which I would prefer to leave it to next PR.

Comment (dentiny) [reply src/ray/core_worker/core_worker.cc]: Implementation updated.

File: src/ray/core_worker/core_worker.h
Comment (dentiny) [reply src/ray/core_worker/core_worker.h]: Sure, release run launched: https://buildkite.com/ray-project/release/builds/27207

Comment (jjyao) [reply src/ray/core_worker/core_worker.h]: Is this with the modification of release test to add env var runtime env?

File: src/ray/core_worker/core_worker.h
Comment (dentiny) [inline src/ray/core_worker/core_worker.h:1870]: @rynewang Caching job runtime env is not easy, since `OverrideRuntimeEnv` takes a job-based runtime env info and **updates** it in-place. I will leave a TODO for it for simplicity (this PR has been back-and-forth for so long), would like to merge it soon.

**Comment by dentiny** [comment]
Release run with no env: https://buildkite.com/ray-project/release/builds/27342

**Comment by dentiny** [comment]
@jjyao Unfortunately still only works for local, and doesn't solve release test performance issue. 
without env: 587.2616526165328
with env: 87.58131870599958

File: src/ray/core_worker/core_worker.h
Comment (rynewang) [inline src/ray/core_worker/core_worker.h:1884]: we can use ThreadSafeSharedLruConstCache to avoid using this mutex here?

File: src/ray/core_worker/core_worker.cc
Comment (rynewang) [inline src/ray/core_worker/core_worker.cc:2205]: I only see this Get, but no Put. Where did you Put()?

Also, on job change we need to reset the cache and I don't see that either. It should be fine because today a core worker is hard wired to a job, and never changes jobs. But if you find it's easy to add that, feel free to add.

Comment (dentiny) [reply src/ray/core_worker/core_worker.cc:2205]: > I only see this Get, but no Put. Where did you Put()?

.... fixed.

Comment (dentiny) [reply src/ray/core_worker/core_worker.cc:2205]: > Also, on job change we need to reset the cache and I don't see that either. It should be fine because today a core worker is hard wired to a job, and never changes jobs. But if you find it's easy to add that, feel free to add.

I'm not familiar with this part, I left a TODO and for next PR, will discuss with you in person.

File: src/ray/core_worker/core_worker.h
Comment (dentiny) [reply src/ray/core_worker/core_worker.h:1887]: No it's intentionally, using mutex guarantees for one particular key (aka, serialized runtime info), only one parsing will happen, so we don't have repeated operation.

Comment (dentiny) [reply src/ray/core_worker/core_worker.h:1887]: To reduce lock contention, I'm planning to implement a sharded LRU with a `GetOrCalculate`-like interface, TODO already left in the shared lru file.

**Comment by dentiny** [comment]
Release build: https://buildkite.com/ray-project/release/builds/27556

**Comment by rynewang** [comment]
DCO sign off

File: src/ray/core_worker/core_worker.h
Comment (rynewang) [reply src/ray/core_worker/core_worker.h:1887]: then I guess we can remove ThreadSafeSharedLruConstCache lol.

**Review by rynewang — APPROVED** [review]

**Comment by dentiny** [comment]
> DCO sign off

Yeah I noticed, but given there're 25 commits in the PR it's little hard and tedious to check which one particular commit fails to sign off. Luckily seems it's not merge blocking.

**Comment by rynewang** [comment]
```
In your local branch, run: git rebase HEAD~25 --signoff
Force push your changes to overwrite the branch: git push --force-with-lease origin hjiang/cache-runtime-env-core-worker
```

**Comment by rynewang** [comment]
I am not a lawyer but if you don't sign DCO there are some legal implication like if the code belongs to you or so, and to avoid future headaches it's better to just rebase before we merge.

**Comment by dentiny** [comment]
> git rebase HEAD~25 --signoff

Yeah that's the same suggestion by chatgpt, but the issue is you don't want to signoff commits, which would unintentionally signoff commits from other people...

**Comment by rynewang** [comment]
then can you use git diff and git apply to make another PR, and I will merge that one right away

**Comment by dentiny** [comment]
> then can you use git diff and git apply to make another PR, and I will merge that one right away

Sure, redo PR: https://github.com/ray-project/ray/pull/49140

**Pull request closed without merging**
